aocd/models.py

```diff
@@ -278,48 +327,70 @@ def title(self):
             raise AocdError(f"unexpected h2 text: {txt}")
         return txt.removeprefix(prefix).removesuffix(suffix)
 
-    def _repr_pretty_(self, p, cycle):
+    def _repr_pretty_(self, p: "IPython.lib.pretty.PrettyPrinter", cycle: bool) -> None:
         """Hook for IPython's pretty-printer."""
-        if cycle:
-            p.text(repr(self))
-        else:
-            txt = f"<Puzzle({self.year}, {self.day}) at {hex(id(self))} - {self.title}>"
-            p.text(txt)
+        txt = repr(self) if cycle else f"<Puzzle({self.year}, {self.day}) at {hex(id(self))} - {self.title}>"
+        p.text(txt) # type: ignore[no-untyped-call] # IPython/Jupyter doesn't provide type hints
 
-    def _coerce_val(self, val):
+    def _coerce_val(self, val: _Answer) -> str:
         # technically adventofcode.com will only accept strings as answers.
         # but it's convenient to be able to submit numbers, since many of the answers
         # are numeric strings. coerce the values to string safely.
-        orig_val = val
-        orig_type = type(val)
+
+        original_val = val
+        def fail() -> NoReturn:
+            msg = f"Failed to coerce {type(original_val).__name__} value {original_val!r} for {self.year}/{self.day:02}."
+            raise AocdError(msg)
+
         coerced = False
-        floatish = isinstance(val, (float, complex))
-        if floatish and val.imag == 0.0 and val.real.is_integer():
-            coerced = True
-            val = int(val.real)
-        elif orig_type.__module__ == "numpy" and getattr(val, "ndim", None) == 0:
-            # deal with numpy scalars
-            if orig_type.__name__.startswith(("int", "uint", "long", "ulong")):
+
+        # if we get an ImportError, numpy is not installed, so we skip handling numpy types
+        with contextlib.suppress(ImportError):
+            import numpy as np
```

Comment:
This is clever - a user can't be submitting a numpy type if numpy isn't avail. I wonder why I didn't think of that originally in [#109](https://github.com/wimglenn/advent-of-code-data/pull/109)  👍